File: 1457-pseudo-palindromic-paths-in-a-binary-tree/1457-pseudo-palindromic-paths-in-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pseudoPalindromicPaths (self, root: Optional[TreeNode]) -> int:
        path = defaultdict(int)
        self.ans = 0
        evens = set()
        
        def dfs(node):
            path[node.val] += 1
            if node.val in evens:
                evens.remove(node.val)
            elif path[node.val] %2 == 0:
                    evens.add(node.val)
            if not(node.left) and not(node.right):
                if len(path) - len(evens) <= 1:
                        self.ans += 1
            else:  
                if node.left: dfs(node.left)
                if node.right: dfs(node.right)
            
            if path[node.val] == 1:
                path.pop(node.val)
            else: path[node.val] -= 1
                
            if node.val in evens:
                evens.remove(node.val)
            elif node.val in path and  path[node.val] %2 == 0:
                evens.add(node.val)
                
        dfs(root)
        return self.ans
            
            
            
                
            

File: 1457-pseudo-palindromic-paths-in-a-binary-tree/1457-pseudo-palindromic-paths-in-a-binary-tree.cpp
/**
 * Definition for a binary tree curr.first.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pseudoPalindromicPaths (TreeNode* root) {        
        stack<pair<TreeNode*,int>> st;      
        st.push({root,0});
        int count = 0;
               
        while (!st.empty()){
            pair<TreeNode*,int> curr =  st.top();
            st.pop();
            if (curr.first != nullptr){
                curr.second =  curr.second ^ (1 << curr.first->val);
                if (curr.first->left == nullptr && curr.first->right == nullptr){
                    if ((curr.second & (curr.second - 1)) == 0){
                        count++;
                    }
                }
                else{
                    st.push({curr.first->right,curr.second});
                    st.push({curr.first->left,curr.second});
                }
            }
        }
        return count;
        
    }
};